## Task 1

### Импортируйте библиотеки pandas и numpy. Загрузите "Boston House Prices dataset" из встроенных наборов данных библиотеки sklearn. 

In [60]:
import numpy as np
import pandas as pd

from sklearn.datasets import load_boston

In [61]:
boston = load_boston()

boston.keys()

dict_keys(['data', 'target', 'feature_names', 'DESCR', 'filename'])

In [62]:
data = boston["data"]

data.shape

(506, 13)

In [63]:
data

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
        4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
        9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
        4.0300e+00],
       ...,
       [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        5.6400e+00],
       [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
        6.4800e+00],
       [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        7.8800e+00]])

In [64]:
feature_names = boston["feature_names"]

feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [65]:
target = boston["target"]

target[:10]

array([24. , 21.6, 34.7, 33.4, 36.2, 28.7, 22.9, 27.1, 16.5, 18.9])

### Создайте датафреймы X и Y из этих данных.
Разбейте эти датафреймы н а тренировочные (X_train, y_train) и тестовые (X_test, y_test) с помощью функции train_test_split так, чтобы размер тестовой выборки
составлял 30% от всех данных, при этом аргумент random_state должен быть равен 42.

In [66]:
X = pd.DataFrame(data, columns=feature_names)

X.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [67]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 13 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
dtypes: float64(13)
memory usage: 51.5 KB


In [150]:
y = pd.DataFrame(target, columns=["price"])

y.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   price   506 non-null    float64
dtypes: float64(1)
memory usage: 4.1 KB


In [151]:
y.head()

price
0   24.0
1   21.6
2   34.7
3   33.4
4   36.2

In [70]:
from sklearn.model_selection import train_test_split

In [71]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

### Создайте модель линейной регрессии под названием lr с помощью класса LinearRegression из модуля sklearn.linear_model.

In [72]:
from sklearn.linear_model import LinearRegression

In [73]:
lr = LinearRegression()

### Обучите модель на тренировочных данных (используйте все признаки) и сделайте предсказание на тестовых.

In [74]:
lr.fit(X_train, y_train)

LinearRegression()

In [75]:
y_pred = lr.predict(X_test)

y_pred.shape

(152, 1)

In [76]:
check_test = pd.DataFrame({
    "y_test": y_test["price"],
    "y_pred": y_pred.flatten(),
})

check_test.head()

y_test     y_pred
173    23.6  28.648960
274    32.4  36.495014
491    13.6  15.411193
72     22.8  25.403213
452    16.1  18.855280

### Вычислите R2 полученных предказаний с помощью r2_score из модуля sklearn.metrics.

In [77]:
from sklearn.metrics import r2_score

r2_score(check_test["y_test"], check_test["y_pred"])

0.7112260057484907

## Task 2

### Создайте модель под названием model с помощью RandomForestRegressor из модуля sklearn.ensemble.¶

In [173]:
from sklearn.ensemble import RandomForestRegressor

In [174]:
from sklearn.model_selection import GridSearchCV

### Сделайте агрумент n_estimators равным 1000, max_depth должен быть равен 12 и random_state сделайте равным 42.

In [191]:
parameters = {
    'n_estimators': [1000],
    'max_depth': [12],
}

clf = GridSearchCV(
    estimator=RandomForestRegressor(random_state=42),
    param_grid=parameters,
    scoring='r2',
    cv=3,
)

### Обучите модель на тренировочных данных аналогично тому, как вы обучали модель
 LinearRegression,
но при этом в метод fit вместо датафрейма y_train поставьте y_train.values[:, 0],
чтобы получить из датафрейма одномерный массив Numpy,
так как для класса RandomForestRegressor в данном методе для аргумента y предпочтительно применение массивов вместо датафрейма.

In [192]:
clf.fit(X_train, y_train.values[:, 0])

GridSearchCV(cv=3, estimator=RandomForestRegressor(random_state=42),
             param_grid={'max_depth': [12], 'n_estimators': [1000]},
             scoring='r2')

cv_results = pd.DataFrame(clf.cv_results_)

cv_results.columns

### Сделайте предсказание на тестовых данных и посчитайте R2. Сравните с результатом из предыдущего задания.

In [194]:
clf = RandomForestRegressor(max_depth=12, n_estimators=1000)

clf.fit(X_train, y_train.values[:, 0])

y_pred = clf.predict(X_test)

r2_score(y_test, y_pred)

0.8727431089344294

### В 1м задании было: 0.7112260057484907

### *Задание 3
Вызовите документацию для класса RandomForestRegressor,
найдите информацию об атрибуте feature_importances_.
С помощью этого атрибута найдите сумму всех показателей важности,
установите, какие два признака показывают наибольшую важность.

In [205]:
importances = clf.feature_importances_

importances

array([0.03496298, 0.00174   , 0.00736248, 0.00111434, 0.01378379,
       0.41159054, 0.01468009, 0.06264408, 0.00491072, 0.01101083,
       0.01741132, 0.01260596, 0.40618286])

In [237]:
sum(importances)

1.0

In [208]:
indices = np.argsort(importances)[::-1]

indices

array([ 5, 12,  7,  0, 10,  6,  4, 11,  9,  2,  8,  1,  3], dtype=int64)

In [236]:
res = list(zip(indices, importances[indices], feature_names[indices]))

res

[(5, 0.4115905449753444, 'RM'),
 (12, 0.4061828557157032, 'LSTAT'),
 (7, 0.06264407906134235, 'DIS'),
 (0, 0.03496298115977857, 'CRIM'),
 (10, 0.017411321213816187, 'PTRATIO'),
 (6, 0.01468008778972864, 'AGE'),
 (4, 0.013783791296873543, 'NOX'),
 (11, 0.012605964326473323, 'B'),
 (9, 0.011010831761937005, 'TAX'),
 (2, 0.007362483208057246, 'INDUS'),
 (8, 0.004910715607540225, 'RAD'),
 (1, 0.0017400046336726776, 'ZN'),
 (3, 0.0011143392497327072, 'CHAS')]

### Самые значимые признаки 5 и 12, т.е. RM и LSTAT